### DISCUSSION TITLE
Definition of realtime for zenoh

### DISCUSSION THREAD

**Discussion by dotChris90** [eclipse-zenoh/roadmap#143] [Q&A] [answered]
May i ask a stupid question related to realtime. 
There exist the repo for bridging zenoh to DDS and seems zenoh is accepted as ROS2 middle ware. There also exist a lot of discussions and programming in area of Eclipse SDV which is automotive area. 

I was now curious because i also saw things like QoS RELIABLE in opposite to zenohs best effort publishing and i also saw the term "REALTIME". But i did'nt see any example in doc or in example folders that could e.g. configure a publisher or subscriber in a way like "hey i have to publish a message in a way that the message arrive to subsriber in 20ms or 200ms" - so basically deadlines. I think they also exist in DDS. 

So is realtime in meaning of "deadlines" a feature that will come or is out already and just undocumented or did i oversee sth? :)

**Comment by kydos** [comment, marked as answer]
Hello @dotChris90,

First of all there are no stupid questions, only stupid answers.

First off, I've just uploaded [a deck](https://bit.ly/4bRRLOj) that should help answer the questions.  Take a look and let me know if this answer your questions.

-- @kydos

**Reply by doganulus** [reply]
@kydos The link says the deck is not found.

**Reply by kydos** [reply]
Seems to be fine now, can you please try again?

**Reply by dotChris90** [reply]
@kydos hello and thanks for the quick reply :D i saw the slides and first was little bit confused because missing deadline listener. on the other side i have to say from the slides your argumentations is reasonable plus the message structures are good enough to check on your own how much time was spend for the transport. on top of that i saw for IPC on the same HW i saw that zenoh enable shared memory which is actually awesome and fastest way you can do to communicate. the execution time then **nearly** should be equal to setting a variable in memory. 

btw i also saw that your messenges have a optional timestamp but according to documentation just for router options - is there a way to enable timestamps also peer to peer? --> https://zenoh.io/docs/manual/abstractions/#timestamp

**Reply by gabrik** [reply]
It should be enough to enable it in the configuration:

```json5
{
  timestamping: {
    enabled: { router: true, peer: true, client: false },
    drop_future_timestamp: false,
  },
}
```

See: https://github.com/eclipse-zenoh/zenoh/blob/main/DEFAULT_CONFIG.json5#L125


If enabled Zenoh expects the peers to have a synchronized time, NTP is enough, otherwise if a machine drift too much message can be dropped (hence  the `drop_future_timestamp`)

**Reply by dotChris90** [reply]
ok got it - but is there a possiblity to do this configuration on code level like 

let mut config = zenoh_config::peer();
config.timestamping.set_enabled(Some(true)).unwrap();

saw it in some test and example folders but here it will not work since the config field enabled: consist of 3 fields { router: true, peer: true, client: false }

**Reply by dotChris90** [reply]
ok i found in zenoh-ext/examples/examples/z_pub_cache.rs how cool! :D 

the code would be 
```rust

use zenoh::config::{Config, ModeDependentValue};

#[tokio::main]
async fn main() {
    let mut config = zenoh_config::peer();
    config.timestamping
          .set_enabled(Some(ModeDependentValue::Unique(true)))
          .unwrap();
    
    let session = zenoh::open(config).res().await.unwrap();
```

so basically zenoh can give timestamps, priority messages, shared memory. 

- Shared memory sounds for me like a extremly nice way to fullfil hard realtime deadlines if message exchange / ipc on the same device 
-  if message exchange / ipc between different ECUs in a wired local environment (not internet) you can set priorities so the real time critical messages find their way first and check timestamps to detect "message delays" and so reasons for not fulfil deadlines 

ok thanks  actually it feels more powerful and still more easy to use than DDS. =)
